ir/memory.cpp

```diff
@@ -195,6 +195,125 @@ ostream& operator<<(ostream &os, const Pointer &p) {
 }
 
 
+Byte::Byte(const Memory &m, expr &&byterepr): m(m), p(move(byterepr)) {
+  assert(p.bits() == m.bitsByte());
+}
+
+Byte::Byte(const Pointer &ptr, unsigned i, expr non_poison)
+    : m(ptr.getMemory()) {
+  // TODO: support pointers larger than 64 bits.
+  assert(m.bitsMemorySize() <= 64 && m.bitsMemorySize() % 8 == 0);
+  // non_poison should be either bool or 1 bit.
+  if (non_poison.isBool())
+    non_poison = non_poison.toBVBool();
+  assert(non_poison.bits() == 1);
+
+  expr byteofs = expr::mkUInt(i, 3);
+  p = expr::mkUInt(1, 1).concat(non_poison).concat(ptr()).concat(byteofs);
+  assert(p.bits() == m.bitsByte());
+}
+
+Byte::Byte(const Memory &m, expr data, expr non_poison): m(m) {
+  assert(m.bitsByte() >= 1 + 8 + 8);
+  assert(data.bits() == 8);
+  assert(non_poison.bits() == 8);
+
+  unsigned padding = m.bitsByte() - 1 - 8 - 8;
+  p = expr::mkUInt(0, 1).concat(expr::mkUInt(0, padding)).concat(non_poison)
+                        .concat(data);
+}
+
+
+void valueToBytes(const StateValue &val, const Type &fromType,
+                  const Memory &mem, vector<Byte> &bytes) {
+  if (fromType.isPtrType()) {
+    Pointer p(mem, val.value);
+    unsigned bytesize = mem.bitsMemorySize() / 8;
+
+    for (unsigned i = 0; i < bytesize; ++i)
+      bytes.emplace_back(p, i, val.non_poison);
+  } else {
+    StateValue bvval = fromType.toBV(val);
+    unsigned bitsize = bvval.bits();
+    unsigned bytesize = divide_up(bitsize, 8);
+
+    bvval = bvval.zext(bytesize * 8 - bitsize);
+
+    for (unsigned i = 0; i < bytesize; ++i) {
+      expr data  = bvval.value.extract((i + 1) * 8 - 1, i * 8);
+      expr np    = bvval.non_poison.extract((i + 1) * 8 - 1, i * 8);
+      bytes.emplace_back(mem, data, np);
+    }
+  }
+}
+
+StateValue bytesToValue(const vector<Byte> &bytes, const Type &toType) {
+  assert(!bytes.empty());
+
+  if (toType.isPtrType()) {
+    bool first = true;
+
+    expr loaded_ptr;
+    // There's no poison byte, and they are all pointer bytes
+    expr all_ptrbytes;
+    // All of the bytes have the same information
+    expr ptr_allequiv;
+    // Byte offsets should be correct
+    expr ptr_correctbyteofs;
+
+    for (unsigned i = 0; i < bytes.size(); ++i) {
+      auto &b = bytes[i];
+      // Is ptr byte?
+      expr is_ptrbyte = b.is_ptrbyte();
+      // Pointer bytes' offset
+      expr byte_ofs = b.ptr_byteoffset();
+      // Pointer's offset
+      expr ptr_value = b.ptr_value();
+      // Is non-poison byte?
+      expr is_nonpoison = b.ptr_nonpoison();
+
+      if (first) {
+        loaded_ptr = move(ptr_value);
+        all_ptrbytes = is_nonpoison & is_ptrbyte;
+        ptr_allequiv = true;
```

Comment:
I think you can use a single non_poison var and keep doing &= on that. Also, if you initialize that, most of this moster case split goes away (there's no need to do SMT simplification here).

Reply:
Makes sense, I updated it and looks better